plutus-playground/plutus-playground-server/app/Auth.hs

```diff
@@ -0,0 +1,315 @@
+                                              setCookieSecure, setCookieValue)
+import qualified Web.JWT                     as JWT
+
+data AuthRole
+  = Anonymous
+  | GithubUser
+  deriving (Show, Eq, Generic, FromJSON, ToJSON)
+
+newtype AuthStatus = AuthStatus
+  { _authStatusAuthRole :: AuthRole
+  } deriving (Show, Eq, Generic, FromJSON, ToJSON)
+
+-- | https://gist.github.com/alpmestan/757094ecf9401f85c5ba367ca20b8900
+type GetRedirect headers = Verb 'GET 302 '[ JSON] (headers NoContent)
+
+-- | We split out the API here because we don't want the Github
+-- Callback to be exported for the frontend to call directly.
+type API
+   = FrontendAPI
+     :<|> CallbackAPI
+
+type FrontendAPI
+   = ("oauth" :> (BrowserHeader "Cookie" Text :> "status" :> Get '[ JSON] AuthStatus
+                  :<|> "github" :> GetRedirect (Headers '[ Header "Location" Text])))
+     :<|> ("gists" :> BrowserHeader "Cookie" Text :> Get '[ JSON] [Gist])
+
+type CallbackAPI
+   = "oauth" :> "github" :> "callback" :> QueryParam "code" OAuthCode :> GetRedirect (Headers '[ Header "Set-Cookie" SetCookie, Header "Location" Text])
+
+data GithubEndpoints = GithubEndpoints
+  { _githubEndpointsAuthLocation        :: !Request
+  , _githubEndpointsAccessTokenLocation :: !Request
+  , _githubEndpointsApiBaseUrl          :: !BaseUrl
+  , _githubEndpointsCallbackUri         :: !Text
+  }
+
+-- | Config determined by Github.
+mkGithubEndpoints :: IO GithubEndpoints
+mkGithubEndpoints = do
+  _githubEndpointsAuthLocation <-
+    parseRequest "GET https://github.com/login/oauth/authorize"
+  _githubEndpointsAccessTokenLocation <-
+    parseRequest "POST https://github.com/login/oauth/access_token"
+  _githubEndpointsApiBaseUrl <- parseBaseUrl "https://api.github.com"
+  let _githubEndpointsCallbackUri = "/api/oauth/github/callback"
+  pure GithubEndpoints {..}
+
+-- | Config supplied at runtime.
+data Config = Config
+  { _configJWTSignature       :: !JWT.Signer
+  , _configRedirectUrl        :: !Text
+  , _configRedirectEndpoint   :: !Text
+  , _configGithubClientId     :: !Text
+  , _configGithubClientSecret :: !Text
+  }
+
+instance FromJSON Config where
+  parseJSON =
+    withObject "config" $ \o -> do
+      _configGithubClientId <- o .: "github-client-id"
+      _configGithubClientSecret <- o .: "github-client-secret"
+      _configJWTSignature <- JWT.hmacSecret <$> o .: "jwt-signature"
+      _configRedirectUrl <- o .: "redirect-url"
+      _configRedirectEndpoint <- o .: "redirect-endpoint"
+      pure Config {..}
+
+hSessionIdCookie :: Text
+hSessionIdCookie = "sessionId"
+
+-- | The JWT key we can lookup for the Github token's value.
+githubTokenClaim :: Text
+githubTokenClaim = "github-token"
+
+redirect ::
+     (Applicative m, ToHttpApiData loc)
+  => loc -- ^ what to put in the 'Location' header
+  -> m (Headers '[ Header "Location" loc] NoContent)
+redirect a = pure $ addHeader a NoContent
```

Comment:
why bother with the `Applicative` here? Is the point of this function to restrict `a` to a `ToHttpApiData`? I just feel it's more general than its name.

Reply:
Good point! `Applicative` is only there because of the `pure` call, but looking afresh there's no need for the `pure` to be there either. 👍